pkg/master/reconcilers/endpointsadapter.go

```diff
@@ -112,8 +112,9 @@ func endpointSliceFromEndpoints(endpoints *corev1.Endpoints) *discovery.Endpoint
 	endpointSlice.Name = endpoints.Name
 	endpointSlice.Labels = map[string]string{discovery.LabelServiceName: endpoints.Name}
 
-	ipAddressType := discovery.AddressTypeIP
-	endpointSlice.AddressType = &ipAddressType
+	// TODO: Add support for IPv6 addresses here (and in the rest of
+	// EndpointsAdapter).
+	endpointSlice.AddressType = discovery.AddressTypeIPv4
```

Comment:
I'm not sold on this - I think IPv4 and IPv6 both "just work" today because we can carry wither on in an Endpoints.  I think you need to iterate the Endpoints and create 1 slice for each family found therein.

@khenidak yes?

Reply:
@thockin Since there's still only one `kubernetes` service created here, and [that service does not specify an IPFamily](https://github.com/kubernetes/kubernetes/blob/master/pkg/master/controller.go#L294-L308), I think it'll default to IPv4. Ideally I think the EndpointSlice would match the Service IPFamily if/until we support multiple Services, one for v4 and one for v6. I think the long term solution involves passing the IPFamily through to the EndpointsAdapter, but that initially felt out of scope here. Would it be sufficient to filter Endpoint addresses here to only add IPv4 addresses [similar to what I'm doing in the EndpointSlice controller here](https://github.com/robscott/kubernetes/blob/endpointslice-iptypes/pkg/controller/endpointslice/https://github.com/robscott/kubernetes/blob/endpointslice-iptypes/pkg/controller/endpointslice/utils.go#L128-L140).

Reply:
I've pushed an update that provides an easier path forward towards true EndpointSlice dual stack support here while also making the code safer for the short term. The current meta proxier kube-proxy implementation will break if an Endpoints resource has more than one IPFamily represented (https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/meta_proxier.go#L177-L203). With that in mind, even if the Endpoints resource here were created with multiple IP families represented, it would just break kube-proxy. I'm assuming the long term solution here involves IP family specific Service, Endpoints, and EndpointSlice, but that feels out of scope here.

For now I've added some safety checks to ensure that if an Endpoints resource does end up with multiple IP families, only IPv4 addresses end up making it into the corresponding EndpointSlice. Additionally, I've added some logic to ensure that EndpointSlices with the deprecated IP address type are replaced with an IPv4 EndpointSlice. This should help for anyone that happens to be upgrading from 1.16 with EndpointSlice alpha enabled (similar functionality already exists in the EndpointSlice controller).

Reply:
>  that service does not specify an IPFamily

@khenidak this sounds like a bug - if the master is on IPv6, the service needs to say so